### ISSUE TITLE
Bug: [no-unnecessary-condition] False positives for Svelte-specific syntax

### ISSUE THREAD

**Issue by mcmxcdev** [typescript-eslint/typescript-eslint#10061]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

Will deliver a repro eventually

### Repro Code

The code below results in `"Unnecessary conditional, value is always falsy."`

```Svelte

export let shouldBeLinked = false

const yesNo = shouldBeLinked ? 'yes' : 'no'
```

This is incorrect because it is a prop that is passed in from the outside and can be `true` as well.

***

The code below results in `"Unnecessary conditional, value is always truthy."`

```Svelte
$: if (locator) {
  resetViewForVariable()
}
```

This is generally correct but Svelte reads the value for changes and retriggers `resetViewForVariable` based on the outcome.

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "'@typescript-eslint/no-unnecessary-condition': 'off',
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

Probably both cases above should be ignored completely by `no-unnecessary-condition`.

### Actual Result

Both cases trigger the `no-unnecessary-condition` and we disable it on our end.

**Comment by bradzacher** [comment]
The rule is powered by the typescript types.
The typescript types say that condition is always truthy / falsey, thus the rule reports.

You need to change the TS types so that they match the runtime semantics of your framework.

**Comment by mcmxcdev** [comment]
>The rule is powered by the typescript types.
>The typescript types say that condition is always truthy / falsey, thus the rule reports.
>
>You need to change the TS types so that they match the runtime semantics of your framework.

The TS types are correct for Svelte usage, thats the problem. 

I was thinking that there is some kind of AST parsing in `typescript-eslint` where you would be able to opt-out of reporting based on certain conditions.

**Comment by bradzacher** [comment]
To be clear - this is why the rule reports - TS sees a `let` variable that is assigned the value `false` and is never reassigned. So it physically sets the type to `false`. So when we inspect it the type is `false`, and hence when used in a condition the rule reports. See this [playground](https://www.typescriptlang.org/play/?#code/KYDwDg9gTgLgBAG2PAzgCwgVwQEwELAAyAlgHYDWwOcAvHAGYCGCKwAUGwMYSkrwCewFADkItOOiy4CJClTgB+OAHJBKZXABcK0hGVsA9AbgnTZ0wD0FHUJFiJkEjNnxEylHACZtAIwgQkRlJxJhZgAG4Obl4BIVFPcUkXGXcqBKVVIQ1tZV19I3NCoqsbcGh4JFRnaTc5HABmEOZWOEYUOD8A4CDIrh4+ODVRRrokmtkPRoy1bJ09Q2MipZMSoA) - even an explicit annotation doesn't override this behaviour. The only way around it is by tricking TS to not see the false via an assertion.

If your framework has some side-effects not represented by the types, then that is not something we can support.

> I was thinking that there is some kind of AST parsing in typescript-eslint where you would be able to opt-out of reporting based on certain conditions.

This exists in a way - an ESLint disable comment.
Otherwise you need to use `false as boolean` as shown above.
